### ISSUE TITLE
Provide Guidance For Provider Authors Writing Tests

### ISSUE THREAD

**Issue by joefitzgerald** [atom/autocomplete-plus#417]
I think we should provide docs :abc: for provider authors to guide them in writing tests for their providers. These tests need to be able to verify a few things:
- [ ] Typing a character triggers a suggestion request
- [ ] The provider is called when their selector matches
- [ ] The provider is not called when their selector does not match
- [ ] [Suggestions](https://github.com/atom-community/autocomplete-plus/wiki/Provider-API#suggestions) are provided when appropriate
- [ ] [Suggestions](https://github.com/atom-community/autocomplete-plus/wiki/Provider-API#suggestions) are not provided when appropriate
- [ ] [Suggestions](https://github.com/atom-community/autocomplete-plus/wiki/Provider-API#suggestions) conform to the autocomplete-plus API definition for suggestions
- [ ] [Suggestions](https://github.com/atom-community/autocomplete-plus/wiki/Provider-API#suggestions) are confirmed correctly
- [ ] The `onDidInsertSuggestion` function (if specified) is called correctly

Further complicating things is that tests:
- Must be asynchronous and leverage `waitsFor`
- Usually need to be integration tests to fully test the desired functionality

Clear and up-to-date documentation for writing testable providers will be invaluable to new provider authors and ensure that the Provider ecosystem continues to expand with good quality, well tested providers.

**Comment by benogle** [comment]
:+1: I like this idea a lot. What about creating a provider repo that showcases these specs? The implementation could be really simple and serve as a template. Copy pasta the code into your new repo, and add your specific functionality.

**Comment by joefitzgerald** [comment]
Yep, that makes sense. Want to create a repo for the autocomplete-example (or similar?) at https://github.com/atom and add me as a contributor?

**Comment by benogle** [comment]
Ok: https://github.com/atom/autocomplete-example You are on the @atom/autocomplete-maintainers team.

**Comment by rm-code** [comment]
> Ok: https://github.com/atom/autocomplete-example You are on the @atom/autocomplete-maintainers team.

I think it would be really nice to have a complete example of how to create a simple autocomplete+ provider.

**Comment by joefitzgerald** [comment]
@rm-code that's the intent of that repo.

**Comment by rm-code** [comment]
@joefitzgerald Oh my bad, I thought it was just there for the specs example. Looking forward to it :+1:

**Comment by joefitzgerald** [comment]
You can't have specs unless there's a provider there to test, also :wink:.

**Comment by DWand** [comment]
It would be interesting to know how to test the case "The provider is not called when their selector does not match".

I've spent about an hour trying to understand why my plugin works fine in the editor, but fails some tests. It looked odd to me. However, I've eventually realized that my plugin is called by autocomplete-plus according to its selector, but in tests I always call it manually (just like it is done in autocomplete-html and autocomplete-css packages) regardless of its selector. How to handle this situation properly? Should I just copy-paste selector matching code from autocomplete-plus or there are some special methods provided for such things?
